File: leetcode/501. Find Mode in Binary Search Tree/main.swift
public class TreeNode {
  public var val: Int
  public var left: TreeNode?
  public var right: TreeNode?
  public init() { self.val = 0; self.left = nil; self.right = nil; }
  public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
  public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
      self.val = val
      self.left = left
      self.right = right
  }
}

class Solution {
    var map: [Int: Int] = [:]
    var max: Int?

    func findMode(_ root: TreeNode?) -> [Int] {
        dfs(root)
        
        var result: [Int] = []
        for (key, value) in map {
            if value == max {
                result.append(key)
            }
        }

        return result
    }

    func dfs(_ node: TreeNode?) {
        guard let node = node else { return }

        dfs(node.left)

        let newValue = (map[node.val] ?? 0) + 1
        map[node.val] = newValue
        if newValue > (max ?? Int.min) {
            max = newValue
        }

        dfs(node.right)
    }
}

File: leetcode/501. Find Mode in Binary Search Tree/main.ts
class TreeNode {
  val: number
  left: TreeNode | null
  right: TreeNode | null
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val
    this.left = left === undefined ? null : left
    this.right = right === undefined ? null : right
  }
}

type OptionalNode = TreeNode | null

const findMode = (root: OptionalNode): number[] => {
    const m = new Map<number, number>()

    const dfs = (node: OptionalNode) => {
        if (!node) {
            return
        }

        dfs(node.left)
        m.set(node.val, (m.get(node.val) ?? 0) + 1)
        dfs(node.right)
    }

    dfs(root)

    const maxValue = Math.max(...m.values())
    const result = Array.from(m.entries()).reduce((prev, item) => {
        if (item[1] === maxValue) {
            return [...prev, item[0]];
        }
        return prev;
    }, []);

    return result
}
